gadopt/energy_solver.py

```diff
@@ -53,82 +50,317 @@
+                    )
+        elif self.mesh.topological_dimension() == 2:
+            self.solver_parameters = direct_energy_solver_parameters.copy()
+        else:
+            self.solver_parameters = iterative_energy_solver_parameters.copy()
+
+        if DEBUG >= log_level:
+            self.solver_parameters["ksp_monitor"] = None
+        elif INFO >= log_level:
+            self.solver_parameters["ksp_converged_reason"] = None
+
+    def su_nubar(self, u: fd.Function, su_diffusivity: float) -> fd.ufl.algebra.Product:
+        """Sets up the streamline-upwind scheme."""
+        if not is_continuous(self.Q):
+            raise TypeError("SU advection requires a continuous function space.")
+
+        log("Using SU advection")
+        # SU(PG) à la Donea & Huerta (2003)
+        # Columns of Jacobian J are the vectors that span the quad/hex and can be
+        # seen as unit vectors scaled with the dx/dy/dz in that direction (assuming
+        # physical coordinates x, y, z aligned with local coordinates).
+        # Thus u^T J is (dx * u , dy * v), and following (2.44c), Pe = u^T J / 2 kappa.
+        # beta(Pe) is the xibar vector in (2.44a)
+        # then we get artifical viscosity nubar from (2.49)
+
+        J = fd.Function(
+            fd.TensorFunctionSpace(self.mesh, "DQ", 1), name="Jacobian"
+        ).interpolate(fd.Jacobian(self.mesh))
+        # Set lower bound for diffusivity in case zero diffusivity specified for
+        # pure advection.
+        kappa = su_diffusivity + 1e-12
+        Pe = absv(fd.dot(u, J)) / 2 / kappa  # Calculate grid Peclet number
+        nubar = su_nubar(u, J, Pe)  # Calculate SU artificial diffusion
+
+        return nubar
+
+    @abc.abstractmethod
+    def set_equation(self):
+        """Sets up the equation."""
+        raise NotImplementedError
+
+    def setup_solver(self) -> None:
+        """Sets up timestepper and associated solver, using specified solver parameters"""
+        self.ts = self.timestepper(
+            self.equation,
+            self.solution,
+            self.delta_t,
+            solution_old=self.solution_old,
+            solver_parameters=self.solver_parameters,
+            strong_bcs=self.strong_bcs,
+        )
+        self._solver_setup = True
+
+    def solve(self, t: Number = 0, update_forcings: Optional[Callable] = None) -> None:
+        """Advances solver in time."""
+        if not self._solver_setup:
+            self.setup_solver()
+
+        self.ts.advance(t, update_forcings)
+
+
+class GenericTransportSolver(GenericTransportBase):
+    """Timestepper and solver for an advection-diffusion equation.
+
+    All combinations of advection, diffusion, sink, and source terms are valid.
+
+    The scalar field is updated in place.
+
+    Arguments:
+      terms:
+        List of equation terms defined in scalar_equation.py
+      scalar_field:
+        Firedrake function for scalar field of interest
+      u:
```

Comment:
Can we make velocity an optional argument of Generic transport solver? I'm thinking of the internal variable timestepping which is just the source and the sink term. 

The energy solver on top would still have it as a required argument but then passes to the Generic transport solver...

Reply:
Yes, sure, that makes sense.